src/main/java/hudson/plugins/promoted_builds/Promotion.java

```diff
@@ -461,32 +463,71 @@ public static ParametersAction getParametersActions(Promotion build){
      * @param actions
      * @param build
      * @param promotionParams
+     * @deprecated Use {@link PromotionParametersAction} with constructor instead.
      */
-	public static void buildParametersAction(@Nonnull List<Action> actions, 
-                @Nonnull AbstractBuild<?, ?> build,
-                @CheckForNull List<ParameterValue> promotionParams) {
-        if (promotionParams == null) {
-            promotionParams = new ArrayList<ParameterValue>();
+    @Deprecated
+    public static void buildParametersAction(@Nonnull List<Action> actions,
+            @Nonnull AbstractBuild<?, ?> build,
+            @CheckForNull List<ParameterValue> promotionParams) {
+        // Create list of actions to pass to scheduled build
+        actions.add(PromotionParametersAction.buildFor(build, promotionParams));
+    }
+
+    private static final Logger LOGGER = Logger.getLogger(Promotion.class.getName());
+    
+    /**
+     * Action, which stores promotion parameters.
+     * This class allows defining custom parameters filtering logic, which is
+     * important for versions after the SECURITY-170 fix.
+     * @since TODO
+     */
+    @Restricted(NoExternalUse.class)
+    public static class PromotionParametersAction extends ParametersAction {
+        
+        private List<ParameterValue> unfilteredParameters;
+        
+        private PromotionParametersAction(List<ParameterValue> params) {
+            // Pass the parameters upstairs
+            super(params);   
```

Comment:
All this parameters are going to be filtered out as they are not defined in the job, right?